server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -556,4 +563,103 @@ public synchronized void cancel() {
             consumeListener();
         }
     }
+
+    /**
+     * Optimized phase result consumer that only counts the number of hits and does not
+     * store any other information.
+     */
+    static class CountOnlyQueryPhaseResultConsumer extends SearchPhaseResults<SearchPhaseResult> {
```

Comment:
This is complex enough I would have it be its own top-level class.

Reply:
I can do that :D